homeassistant/components/enocean/binary_sensor.py

```diff
@@ -1,56 +1,123 @@
 """Support for EnOcean binary sensors."""
 import logging
+from typing import List, Optional
 
 import voluptuous as vol
 
-from homeassistant.components import enocean
 from homeassistant.components.binary_sensor import (
     DEVICE_CLASSES_SCHEMA,
     PLATFORM_SCHEMA,
     BinarySensorEntity,
 )
 from homeassistant.const import CONF_DEVICE_CLASS, CONF_ID, CONF_NAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import entity_registry
 import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_registry import async_entries_for_config_entry
+
+from .const import DATA_ENOCEAN, DOMAIN
+from .device import EnOceanEntity
+from .utilities import enocean_id_to_string, string_to_enocean_id
 
 _LOGGER = logging.getLogger(__name__)
 
 DEFAULT_NAME = "EnOcean binary sensor"
 DEPENDENCIES = ["enocean"]
 EVENT_BUTTON_PRESSED = "button_pressed"
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {
-        vol.Required(CONF_ID): vol.All(cv.ensure_list, [vol.Coerce(int)]),
-        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
-    }
+PLATFORM_SCHEMA = vol.All(
+    cv.deprecated(CONF_ID),
+    PLATFORM_SCHEMA.extend(
+        {
+            vol.Required(CONF_ID): vol.All(cv.ensure_list, [vol.Coerce(int)]),
+            vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+            vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+        },
+    ),
 )
 
+YAML_CONFIG_ENTRIES_TO_IMPORT = "config_entries_to_import"
+
+
+async def async_setup_platform(hass, config, add_entities, discovery_info=None):
+    """Convert the config entry to an entity in the registry."""
+    if not hass.config_entries.async_entries(domain=DOMAIN):
+        return
+
+    yaml_entries_to_migrate = hass.data[DATA_ENOCEAN].setdefault(
+        YAML_CONFIG_ENTRIES_TO_IMPORT, []
+    )
+    yaml_entries_to_migrate.append(config)
+
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the Binary Sensor platform for EnOcean."""
-    dev_id = config.get(CONF_ID)
-    dev_name = config.get(CONF_NAME)
-    device_class = config.get(CONF_DEVICE_CLASS)
+async def async_setup_entry(hass: HomeAssistant, config_entry, async_add_entities):
+    """Set up binary sensors for the ENOcean component."""
 
-    add_entities([EnOceanBinarySensor(dev_id, dev_name, device_class)])
+    entities_to_add = []
 
+    # Load entities from registry
+    entities_registry = await entity_registry.async_get_registry(hass)
+    entries_from_config = async_entries_for_config_entry(
+        entities_registry, config_entry.entry_id
+    )
+    for entry in entries_from_config:
+        entities_to_add.append(
+            EnOceanBinarySensor(
```

Comment:
How will new entities be added for new users?

Reply:
With this PR, the YAML config is still the only way to declare entities.
My initial objective was to do that work in the scope of another PR, to limit the size of this one (mostly focusing on migrating the integration to flows).

The EnOcean protocol processing logic will require a major change in the next steps, notably to be able to associate devices automatically. Some steps for those changes would be:
- rename the current entities to legacy, and develop new entities, that relies on the device EEP to decode and create EnOcean packets
- update the code of the dongle to get the dongle device ID at startup
- adding code to handle teach-in packets, that will allow the user to add entities by using the pairing procedure of their devices if available
- adding code to handle data packets with learn bit set, and ask the EEP of the device to the user if needed

I have prepared most of the logic for those steps. Do you think I should add it to this PR, to have a fully functional integration in the end ?

Reply:
It's a bit weird that we deprecate the platform config yaml when it's needed to add new devices. Maybe the order of these PRs should be different?

A good thing is that we cut the beta last Wednesday, so there's 2.5 weeks until next beta. Will that be enough?

I don't think we should add more code to this PR.

Reply:
Note that manual configuration will always be a thing for EnOcean. Quite a lot of the most common devices do not support pairing and and auto-detection based on the payload is often not possible

Reply:
Then I'm not sure we can merge this until there's a way to add new devices manually.

Reply:
Now that I think of, automatically detecting the _specific_ device type without manual intervention is not possible, because the device groups are too big / diverse.
If you want to go big, you certainly can detect new devices and then somehow have the user select the specific device type in the UI. That certainly would be cool (and a lot of work). I just wouldn't deprecate the manual configuration option before that is in place.

Reply:
Ok, I think we can make this work if we use the same approach as the mqtt integration does. Have a config entry for the connection, and then config yaml for each platform.

Let's remove the migration for now.

Reply:
In a prototype I have built to learn before proposing a PR, I was using a config flow to ask the user for a the exact type of the device. The flow was aborted at the end, after a message was sent to request the creation of the entity.
It was definitely hacky, but it was working.

I will stash the migration code tomorrow.